In [1]:
import pandas as pd

melb_df = pd.read_csv('data/melb_data_fe.csv')
melb_df.head()

Suburb  Rooms   Type      Price Method SellerG        Date  Distance  \
0  Abbotsford      2  house  1480000.0      S  Biggin  2016-03-12       2.5   
1  Abbotsford      2  house  1035000.0      S  Biggin  2016-04-02       2.5   
2  Abbotsford      3  house  1465000.0     SP  Biggin  2017-04-03       2.5   
3  Abbotsford      3  house   850000.0     PI  Biggin  2017-04-03       2.5   
4  Abbotsford      4  house  1600000.0     VB  Nelson  2016-04-06       2.5   

   Postcode  Bedroom  ...  Longtitude             Regionname  Propertycount  \
0      3067        2  ...    144.9984  Northern Metropolitan           4019   
1      3067        2  ...    144.9934  Northern Metropolitan           4019   
2      3067        3  ...    144.9944  Northern Metropolitan           4019   
3      3067        3  ...    144.9969  Northern Metropolitan           4019   
4      3067        3  ...    144.9941  Northern Metropolitan           4019   

   MeanRoomsSquare AreaRatio  MonthSale  AgeBuilding WeekdaySale  StreetType  \
0            25.20 -0.231707          3           46           5          St   
1            15.80 -0.327660          4          116           5          St   
2            18.75  0.056338          4          117           0          St   
3            15.75  0.145455          4           47           0       other   
4            17.75  0.083969          4            2           2          St   

   Weekend  
0        1  
1        1  
2        0  
3        0  
4        0  

[5 rows x 26 columns]

In [2]:
display(melb_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 26 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Suburb           13580 non-null  object 
 1   Rooms            13580 non-null  int64  
 2   Type             13580 non-null  object 
 3   Price            13580 non-null  float64
 4   Method           13580 non-null  object 
 5   SellerG          13580 non-null  object 
 6   Date             13580 non-null  object 
 7   Distance         13580 non-null  float64
 8   Postcode         13580 non-null  int64  
 9   Bedroom          13580 non-null  int64  
 10  Bathroom         13580 non-null  int64  
 11  Car              13580 non-null  int64  
 12  Landsize         13580 non-null  float64
 13  BuildingArea     13580 non-null  float64
 14  CouncilArea      12211 non-null  object 
 15  Lattitude        13580 non-null  float64
 16  Longtitude       13580 non-null  float64
 17  Regionname  

None

In [4]:

melb_df['Date'] = pd.to_datetime(melb_df['Date'], dayfirst=True)

In [5]:
melb_df['Date'].dt.quarter.value_counts()

3    4873
2    4359
4    2329
1    2019
Name: Date, dtype: int64

In [6]:
melb_df.nunique()

Suburb              120
Rooms                 9
Type                  3
Price              2204
Method                5
SellerG              50
Date                 58
Distance            202
Postcode            198
Bedroom              12
Bathroom              9
Car                  11
Landsize           1448
BuildingArea        602
CouncilArea          33
Lattitude          6503
Longtitude         7063
Regionname            8
Propertycount       311
MeanRoomsSquare    1275
AreaRatio          6521
MonthSale            12
AgeBuilding         152
WeekdaySale           7
StreetType           11
Weekend               2
dtype: int64

In [7]:
melb_df.sort_values(by='Price').head(10)

Suburb  Rooms   Type     Price Method        SellerG  \
2652        Footscray      1   unit   85000.0     PI        Burnham   
1805            other      4  house  131000.0     PI          other   
7303           Albion      1   unit  145000.0     PI         Biggin   
1927           Coburg      4  house  145000.0     PI         Jellis   
7940         Hawthorn      1   unit  160000.0     VB            HAR   
12666       Brunswick      1   unit  170000.0     VB         Nelson   
8811        Footscray      1   unit  170000.0     PI        Burnham   
8504   West Footscray      1   unit  185000.0     PI            Jas   
7293           Albion      1   unit  185000.0      S  hockingstuart   
7305           Albion      2   unit  190000.0     SP        Burnham   

            Date  Distance  Postcode  Bedroom  ...  Longtitude  \
2652  2016-03-09       6.4      3011        1  ...   144.89000   
1805  2017-02-25       8.9      3162        4  ...   145.02420   
7303  2016-05-28      13.9      3020        2  ...   144.82660   
1927  2016-04-06       7.8      3058        3  ...   144.96580   
7940  2017-08-04       4.6      3122        1  ...   145.03730   
12666 2017-09-16       5.2      3056        1  ...   144.95188   
8811  2017-01-07       5.1      3011        1  ...   144.89587   
8504  2017-04-29       8.2      3012        1  ...   144.86720   
7293  2016-08-22      13.9      3020        1  ...   144.82720   
7305  2016-07-30      13.9      3020        2  ...   144.82390   

                  Regionname  Propertycount  MeanRoomsSquare AreaRatio  \
2652    Western Metropolitan           7570        42.000000  1.000000   
1805   Southern Metropolitan           2379        17.222222 -0.525994   
7303    Western Metropolitan           2185        31.500000  0.555556   
1927   Northern Metropolitan          11204        20.500000 -0.531429   
7940   Southern Metropolitan          11308        42.000000 -0.437500   
12666  Northern Metropolitan          11918        42.000000 -0.816860   
8811    Western Metropolitan           7570         8.666667 -0.071429   
8504    Western Metropolitan           5058        42.000000  1.000000   
7293    Western Metropolitan           2185        14.333333 -0.964017   
7305    Western Metropolitan           2185        25.200000  1.000000   

       MonthSale  AgeBuilding WeekdaySale  StreetType  Weekend  
2652           3            9           2          St        0  
1805           2           97           5          St        1  
7303           5           46           5          St        1  
1927           4          106           2          Rd        0  
7940           8            8           4          St        0  
12666          9           47           5          St        1  
8811           1            4           5          St        1  
8504           4           47           5          St        1  
7293           8           41           0          Rd        0  
7305           7           46           5          St        1  

[10 rows x 26 columns]

In [8]:
melb_df.sort_values(by='Date', ascending=False)

Suburb  Rooms   Type      Price Method    SellerG       Date  \
11144     Northcote      4  house  1955000.0     SP    McGrath 2017-12-08   
11217  Surrey Hills      3  house  1775000.0     PI     Jellis 2017-12-08   
11206      St Kilda      4  house  1600000.0     VB       Gary 2017-12-08   
11207    Strathmore      7  house  2000000.0     SP     Nelson 2017-12-08   
11208    Strathmore      4  house  1610000.0      S      other 2017-12-08   
...             ...    ...    ...        ...    ...        ...        ...   
1555     Camberwell      4  house  2650000.0      S     Jellis 2016-03-09   
4510       Oak Park      3  house   892000.0      S     Nelson 2016-03-09   
4388      Northcote      3  house  1200000.0      S     Nelson 2016-03-09   
6184   Surrey Hills      3  house  1205000.0      S  Fletchers 2016-01-28   
6183   Surrey Hills      2   unit   813000.0      S  Fletchers 2016-01-28   

       Distance  Postcode  Bedroom  ...  Longtitude             Regionname  \
11144       5.3      3070        4  ...   144.99375  Northern Metropolitan   
11217      10.2      3127        3  ...   145.08968  Southern Metropolitan   
11206       5.0      3182        4  ...   144.98324  Southern Metropolitan   
11207       8.2      3041        7  ...   144.90587   Western Metropolitan   
11208       8.2      3041        4  ...   144.92243   Western Metropolitan   
...         ...       ...      ...  ...         ...                    ...   
1555        7.8      3124        4  ...   145.06860  Southern Metropolitan   
4510       11.5      3046        3  ...   144.91740  Northern Metropolitan   
4388        5.5      3070        3  ...   144.99090  Northern Metropolitan   
6184       11.2      3127        3  ...   145.10060  Southern Metropolitan   
6183       11.2      3127        2  ...   145.10230  Southern Metropolitan   

       Propertycount  MeanRoomsSquare AreaRatio  MonthSale  AgeBuilding  \
11144          11364        12.600000  0.923664         12           47   
11217           5457        15.750000 -0.721854         12           47   
11206          13240        16.000000 -0.586028         12          107   
11207           3284        20.882353 -0.262721         12           18   
11208           3284        24.909091 -0.517181         12           62   
...              ...              ...       ...        ...          ...   
1555            8920        18.900000 -0.550535          3          126   
4510            2651        18.000000 -0.714609          3           46   
4388           11364        17.857143 -0.264706          3           96   
6184            5457        18.000000 -0.590909          1           46   
6183            5457        25.200000  0.076923          1           46   

      WeekdaySale  StreetType  Weekend  
11144           4          St        0  
11217           4          Rd        0  
11206           4          St        0  
11207           4          Rd        0  
11208           4          St        0  
...           ...         ...      ...  
1555            2          Rd        0  
4510            2          St        0  
4388            2          St        0  
6184            3          Rd        0  
6183            3          Rd        0  

[13580 rows x 26 columns]

In [9]:
melb_df.sort_values(by=['Distance', 'Price']).loc[::10, ['Distance', 'Price']]

Distance      Price
11428       0.0   387000.0
10512       0.7   600000.0
5727        1.2   485000.0
8671        1.2   595000.0
5736        1.2   740000.0
...         ...        ...
12011      38.0   680000.0
10673      38.0   810000.0
13429      38.0  1155000.0
11102      41.0   650000.0
12728      45.9   650000.0

[1358 rows x 2 columns]

In [11]:
cols_to_exclude = ['Date', 'Rooms', 'Bedroom', 'Bathroom', 'Car'] 
max_unique_count = 150 
for col in melb_df.columns: 
    if melb_df[col].nunique() < max_unique_count and col not in cols_to_exclude: 
        melb_df[col] = melb_df[col].astype('category')
display(melb_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 26 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   Suburb           13580 non-null  category      
 1   Rooms            13580 non-null  int64         
 2   Type             13580 non-null  category      
 3   Price            13580 non-null  float64       
 4   Method           13580 non-null  category      
 5   SellerG          13580 non-null  category      
 6   Date             13580 non-null  datetime64[ns]
 7   Distance         13580 non-null  float64       
 8   Postcode         13580 non-null  int64         
 9   Bedroom          13580 non-null  int64         
 10  Bathroom         13580 non-null  int64         
 11  Car              13580 non-null  int64         
 12  Landsize         13580 non-null  float64       
 13  BuildingArea     13580 non-null  float64       
 14  CouncilArea      12211 non-null  categ

None

In [10]:
mask1 = melb_df['AreaRatio'] < -0.8
mask2 = melb_df['Type'] == 'townhouse'
mask3 = melb_df['SellerG'] == 'McGrath'
melb_df[mask1 & mask2 & mask3].sort_values(
    by=['Date', 'AreaRatio'],
    ascending=[True, False],
    ignore_index=True
).loc[:, ['Date', 'AreaRatio']]

Date  AreaRatio
0 2016-07-26  -0.974922
1 2016-09-24  -0.971831
2 2016-11-27  -0.953608
3 2016-12-11  -0.945946
4 2017-08-04  -0.947368
5 2017-08-04  -0.970874

In [12]:
int(melb_df.sort_values(
    by='AreaRatio', 
    ignore_index=True,
    ascending=False
).loc[1558, 'BuildingArea'])

126

In [16]:
mask1 = melb_df['Rooms'] > 2
mask2 = melb_df['Type'] == 'townhouse'
melb_df[mask1 & mask2].sort_values(
    by=['Rooms', 'MeanRoomsSquare'],
    ascending=[True, False],
    ignore_index=True
).loc[18, ['Rooms', 'MeanRoomsSquare','Price']]

Rooms                      3
MeanRoomsSquare        27.75
Price              1300000.0
Name: 18, dtype: object

In [17]:
melb_df.groupby(by='Type').mean()

Rooms         Price   Distance     Postcode   Bedroom  Bathroom  \
Type                                                                            
house      3.260874  1.242665e+06  10.979479  3104.080643  3.229336  1.613822   
townhouse  2.837522  9.337351e+05   9.851346  3100.777379  2.814183  1.809695   
unit       1.963871  6.051275e+05   7.607391  3110.797481  1.966523  1.183295   

                Car    Landsize  BuildingArea  Lattitude  Longtitude  \
Type                                                                   
house      1.772674  617.181924    152.162553 -37.803795  144.994700   
townhouse  1.555655  279.606822    134.649710 -37.815782  144.996489   
unit       1.128936  477.314219    102.235863 -37.823710  144.996363   

           Propertycount  MeanRoomsSquare  AreaRatio  AgeBuilding  
Type                                                               
house        7259.025505        18.996731  -0.490031    55.669700  
townhouse    7094.459605        18.569847  -0.094916    26.690305  
unit         8199.280080        21.068242   0.319883    39.703016

In [18]:
melb_df.groupby('Type')['Price'].mean()

Type
house        1.242665e+06
townhouse    9.337351e+05
unit         6.051275e+05
Name: Price, dtype: float64

In [19]:
melb_df.groupby('Regionname')['Distance'].min().sort_values(ascending=False)

Regionname
Western Victoria              29.8
Eastern Victoria              25.2
Northern Victoria             21.8
South-Eastern Metropolitan    14.7
Eastern Metropolitan           7.8
Western Metropolitan           4.3
Southern Metropolitan          0.7
Northern Metropolitan          0.0
Name: Distance, dtype: float64

In [20]:
melb_df.groupby('MonthSale')['Price'].agg(
    ['count', 'mean', 'max']
).sort_values(by='count', ascending=False)

count          mean        max
MonthSale                                
8           1850  1.056371e+06  6500000.0
7           1835  9.314698e+05  9000000.0
5           1644  1.097807e+06  8000000.0
6           1469  1.068981e+06  7650000.0
3           1408  1.146762e+06  5600000.0
4           1246  1.050479e+06  5500000.0
9           1188  1.126349e+06  6400000.0
10           854  1.135970e+06  6250000.0
11           750  1.142503e+06  5050000.0
12           725  1.144737e+06  5700000.0
2            333  1.169051e+06  4735000.0
1            278  9.397921e+05  5200000.0

In [21]:
melb_df.groupby('MonthSale')['Price'].agg('describe')

count          mean            std       min       25%        50%  \
MonthSale                                                                       
1           278.0  9.397921e+05  577668.924214  170000.0  570500.0   795000.0   
2           333.0  1.169051e+06  671564.357417  131000.0  710000.0  1020000.0   
3          1408.0  1.146762e+06  709573.596867   85000.0  680000.0   945000.0   
4          1246.0  1.050479e+06  591892.902979  145000.0  655000.0   905500.0   
5          1644.0  1.097807e+06  668492.867996  145000.0  650000.0   905000.0   
6          1469.0  1.068981e+06  606010.069052  222000.0  660000.0   900000.0   
7          1835.0  9.314698e+05  537390.803161  190000.0  586750.0   800000.0   
8          1850.0  1.056371e+06  619617.476541  160000.0  635000.0   892000.0   
9          1188.0  1.126349e+06  608734.690742  170000.0  725000.0   980000.0   
10          854.0  1.135970e+06  692950.251627  250000.0  652625.0   950000.0   
11          750.0  1.142503e+06  703666.227892  240000.0  671750.0   950000.0   
12          725.0  1.144737e+06  706549.007161  250000.0  667500.0   938000.0   

                 75%        max  
MonthSale                        
1          1111250.0  5200000.0  
2          1478000.0  4735000.0  
3          1400000.0  5600000.0  
4          1298750.0  5500000.0  
5          1371250.0  8000000.0  
6          1325000.0  7650000.0  
7          1150000.0  9000000.0  
8          1310000.0  6500000.0  
9          1360000.0  6400000.0  
10         1416500.0  6250000.0  
11         1399125.0  5050000.0  
12         1412000.0  5700000.0

In [22]:
melb_df.groupby('Regionname')['SellerG'].agg(
    		['nunique', set]
)

nunique  \
Regionname                            
Eastern Metropolitan             26   
Eastern Victoria                 11   
Northern Metropolitan            40   
Northern Victoria                11   
South-Eastern Metropolitan       25   
Southern Metropolitan            38   
Western Metropolitan             34   
Western Victoria                  6   

                                                                          set  
Regionname                                                                     
Eastern Metropolitan        {Jellis, RW, Barry, Nelson, RT, hockingstuart,...  
Eastern Victoria            {Harcourts, Eview, other, Barry, McGrath, O'Br...  
Northern Metropolitan       {Jas, Eview, Jellis, Nelson, Barry, RW, RT, ho...  
Northern Victoria           {LITTLE, other, Barry, McGrath, Raine, Bucking...  
South-Eastern Metropolitan  {Eview, Jellis, Nelson, Barry, RW, O'Brien, ho...  
Southern Metropolitan       {Jas, Eview, Jellis, RW, RT, Barry, Nelson, O'...  
Western Metropolitan        {Jas, Jellis, Nelson, Barry, RT, RW, O'Brien, ...  
Western Victoria                 {other, Raine, hockingstuart, Ray, YPA, HAR}

In [23]:
melb_df.groupby('Rooms')['Price'].mean()

Rooms
1     4.338245e+05
2     7.750812e+05
3     1.076081e+06
4     1.445282e+06
5     1.870260e+06
6     1.849366e+06
7     1.920700e+06
8     1.602750e+06
10    9.000000e+05
Name: Price, dtype: float64

In [25]:
melb_df.groupby('Regionname')['Lattitude'].agg(
    ['count', 'mean', 'std']
).sort_values(by='std', ascending=False)

count       mean       std
Regionname                                            
Eastern Victoria               53 -37.959292  0.147067
Northern Victoria              41 -37.599620  0.084455
South-Eastern Metropolitan    450 -37.993727  0.073411
Western Metropolitan         2948 -37.782382  0.051251
Northern Metropolitan        3890 -37.747258  0.049639
Eastern Metropolitan         1471 -37.788820  0.047890
Southern Metropolitan        4695 -37.866981  0.043080
Western Victoria               32 -37.695163  0.011579

In [26]:
melb_df['Date'] = pd.to_datetime(melb_df['Date'], dayfirst=True)

In [28]:
date1 = pd.to_datetime('2017-05-01')
date2 = pd.to_datetime('2017-09-01')
mask = (date1 <= melb_df['Date']) & (melb_df['Date']<= date2)
melb_df[mask].groupby('SellerG')['Price'].sum().sort_values(ascending=True)

SellerG
LITTLE             2742000.0
Cayzer             4439000.0
Burnham            4550500.0
Moonee             7328000.0
Thomson            8332000.0
Bells              8656000.0
Alexkarbon        10985000.0
McDonald          14637500.0
Rendina           15422276.0
Nick              16890000.0
Douglas           18341000.0
Buckingham        19033000.0
C21               19515000.0
Eview             19791500.0
Collins           20217000.0
Philip            22051800.0
Chisholm          23225000.0
Williams          23297000.0
Love              23365500.0
Purplebricks      23401000.0
O'Brien           23855508.0
HAR               25568000.0
Village           26473000.0
RW                29261000.0
Raine             30687700.0
Stockdale         35409800.0
Sweeney           36882750.0
Gary              39138400.0
Hodges            43231000.0
YPA               46354350.0
Miles             47582000.0
Kay               48569500.0
RT                50498000.0
Brad              55955000.0
Jas   

In [29]:
melb_df.groupby('Rooms')[['Price', 'BuildingArea']].median()

Price  BuildingArea
Rooms                         
1       385000.0         107.0
2       690000.0         126.0
3       950000.0         126.0
4      1285000.0         142.0
5      1660000.0         176.0
6      1800000.0         126.0
7      1496000.0         216.5
8      1515000.0         126.0
10      900000.0         126.0

In [30]:
melb_df.groupby(['Rooms', 'Type'])['Price'].mean()

Rooms  Type     
1      house        8.668655e+05
       townhouse    5.927045e+05
       unit         3.899289e+05
2      house        1.017238e+06
       townhouse    7.101585e+05
       unit         6.104905e+05
3      house        1.109233e+06
       townhouse    9.847087e+05
       unit         8.505963e+05
4      house        1.462283e+06
       townhouse    1.217092e+06
       unit         1.037476e+06
5      house        1.877327e+06
       townhouse    1.035000e+06
       unit                  NaN
6      house        1.869508e+06
       townhouse             NaN
       unit         5.200000e+05
7      house        1.920700e+06
       townhouse             NaN
       unit                  NaN
8      house        1.510286e+06
       townhouse             NaN
       unit         2.250000e+06
10     house        9.000000e+05
       townhouse             NaN
       unit                  NaN
Name: Price, dtype: float64

In [31]:
melb_df.groupby(['Rooms', 'Type'])['Price'].mean().unstack()

Type          house     townhouse          unit
Rooms                                          
1      8.668655e+05  5.927045e+05  3.899289e+05
2      1.017238e+06  7.101585e+05  6.104905e+05
3      1.109233e+06  9.847087e+05  8.505963e+05
4      1.462283e+06  1.217092e+06  1.037476e+06
5      1.877327e+06  1.035000e+06           NaN
6      1.869508e+06           NaN  5.200000e+05
7      1.920700e+06           NaN           NaN
8      1.510286e+06           NaN  2.250000e+06
10     9.000000e+05           NaN           NaN

In [32]:
melb_df.pivot_table(
    values='Price',
    index='Rooms',
    columns='Type',
    fill_value=0
).round()

Type       house  townhouse       unit
Rooms                                 
1       866866.0   592705.0   389929.0
2      1017238.0   710158.0   610491.0
3      1109233.0   984709.0   850596.0
4      1462283.0  1217092.0  1037476.0
5      1877327.0  1035000.0        0.0
6      1869508.0        0.0   520000.0
7      1920700.0        0.0        0.0
8      1510286.0        0.0  2250000.0
10      900000.0        0.0        0.0

In [33]:
melb_df.pivot_table(
    values='Price',
    index='Regionname',
    columns='Weekend',
    aggfunc='count',
)

Weekend                        0     1
Regionname                            
Eastern Metropolitan         447  1024
Eastern Victoria              13    40
Northern Metropolitan       1258  2632
Northern Victoria             11    30
South-Eastern Metropolitan   123   327
Southern Metropolitan       1534  3161
Western Metropolitan         960  1988
Western Victoria               8    24

In [34]:
melb_df.pivot_table(
    values='Landsize',
    index='Regionname',
    columns='Type',
    aggfunc=['median', 'mean'],
    fill_value=0
)

median                        mean              \
Type                        house townhouse unit        house   townhouse   
Regionname                                                                  
Eastern Metropolitan        674.0     233.5  203   717.422847  269.440678   
Eastern Victoria            843.0       0.0  230  3108.960000    0.000000   
Northern Metropolitan       459.5     134.0    0   619.249092  317.325733   
Northern Victoria           724.0       0.0    0  3355.463415    0.000000   
South-Eastern Metropolitan  630.5     240.0  199   664.306701  212.160000   
Southern Metropolitan       586.0     246.0    0   569.643881  278.858824   
Western Metropolitan        531.0     198.0   62   507.883406  244.560669   
Western Victoria            599.5       0.0    0   655.500000    0.000000   

                                        
Type                              unit  
Regionname                              
Eastern Metropolitan        330.444444  
Eastern Victoria            295.333333  
Northern Metropolitan       495.026538  
Northern Victoria             0.000000  
South-Eastern Metropolitan  357.864865  
Southern Metropolitan       466.380245  
Western Metropolitan        557.637232  
Western Victoria              0.000000

In [35]:
melb_df.pivot_table(
    values='Price',
    index=['Method','Type'],
    columns='Regionname',
    aggfunc='median',
    fill_value=0
)

Regionname        Eastern Metropolitan  Eastern Victoria  \
Method Type                                                
PI     house                   1244000            780000   
       townhouse                760000                 0   
       unit                     650000                 0   
S      house                   1127000            675000   
       townhouse                828000                 0   
       unit                     645750            492000   
SA     house                    932500            950000   
       townhouse                807500                 0   
       unit                          0                 0   
SP     house                   1050000            672500   
       townhouse                910000                 0   
       unit                     515000            400000   
VB     house                   1100000            712500   
       townhouse                892500                 0   
       unit                     500000                 0   

Regionname        Northern Metropolitan  Northern Victoria  \
Method Type                                                  
PI     house                     900000             500000   
       townhouse                 632500                  0   
       unit                      410000                  0   
S      house                     920000             555000   
       townhouse                 750000                  0   
       unit                      525500                  0   
SA     house                     817500             540000   
       townhouse                 425000                  0   
       unit                      616000                  0   
SP     house                     900000             521000   
       townhouse                 690000                  0   
       unit                      470000                  0   
VB     house                    1050000             690000   
       townhouse                 640000                  0   
       unit                      450000                  0   

Regionname        South-Eastern Metropolitan  Southern Metropolitan  \
Method Type                                                           
PI     house                          865000                1725000   
       townhouse                     1190000                1055000   
       unit                           525000                 571250   
S      house                          883300                1611000   
       townhouse                      875000                1135000   
       unit                           606000                 655000   
SA     house                          880000                1390000   
       townhouse                           0                1141000   
       unit                                0                 580000   
SP     house                          770000                1521750   
       townhouse                      800000                1162500   
       unit                           601000                 550000   
VB     house                          850000                1800000   
       townhouse                           0                1250000   
       unit                           700000                 500000   

Regionname        Western Metropolitan  Western Victoria  
Method Type                                               
PI     house                    870000            630000  
       townhouse                670000                 0  
       unit                     360000                 0  
S      house                    870000            397500  
       townhouse                729000                 0  
       unit                     489000                 0  
SA     house                    772500                 0  
       townhouse                467500                 0  
       unit                     571000                 0  
SP     house                    865000            360000  
       t

In [36]:
pivot = melb_df.pivot_table(
    values='Landsize',
    index='Regionname',
    columns='Type',
    aggfunc=['median', 'mean'],
    fill_value=0
)

In [37]:
pivot.columns

MultiIndex([('median',     'house'),
            ('median', 'townhouse'),
            ('median',      'unit'),
            (  'mean',     'house'),
            (  'mean', 'townhouse'),
            (  'mean',      'unit')],
           names=[None, 'Type'])

In [38]:
display(pivot['mean']['unit'])

Regionname
Eastern Metropolitan          330.444444
Eastern Victoria              295.333333
Northern Metropolitan         495.026538
Northern Victoria               0.000000
South-Eastern Metropolitan    357.864865
Southern Metropolitan         466.380245
Western Metropolitan          557.637232
Western Victoria                0.000000
Name: unit, dtype: float64

In [39]:
mask = pivot['mean']['house'] < pivot['median']['house']
filtered_pivot = pivot[mask]
display(filtered_pivot)

median                       mean              \
Type                   house townhouse unit       house   townhouse   
Regionname                                                            
Southern Metropolitan  586.0     246.0    0  569.643881  278.858824   
Western Metropolitan   531.0     198.0   62  507.883406  244.560669   

                                   
Type                         unit  
Regionname                         
Southern Metropolitan  466.380245  
Western Metropolitan   557.637232

In [43]:
melb_df.pivot_table(
    values='Price',
    index='SellerG',
    columns='Type',
    aggfunc='median',
).sort_values('unit')

Type               house  townhouse      unit
SellerG                                      
Burnham         922500.0        NaN  305000.0
Sweeney         850000.0   728000.0  381000.0
Stockdale       715500.0   550000.0  410000.0
Village         973500.0   727500.0  420000.0
Love            780000.0   846000.0  440000.0
YPA             596250.0   498500.0  443500.0
LITTLE          970000.0   800000.0  447000.0
Raine           745000.0   608000.0  447500.0
Jas             935000.0   699000.0  455000.0
Bells           692500.0   430500.0  457500.0
Rendina        1190000.0   750000.0  459000.0
Brad            875000.0   700000.0  465000.0
Douglas         684000.0   500000.0  494500.0
Thomson        1177000.0   500000.0  495000.0
Ray             900000.0   650000.0  497500.0
Harcourts       740000.0   653750.0  500000.0
Barry           830000.0   727500.0  502000.0
McDonald       1242500.0        NaN  502500.0
HAR             695000.0        NaN  518000.0
Nelson         1055000.0   730000.0  530000.0
other           901000.0   840500.0  535000.0
McGrath        1250000.0   810000.0  542500.0
Greg           1330000.0   900000.0  547500.0
Biggin         1151750.0   675000.0  548000.0
Gary           1370000.0   993500.0  550000.0
Buckingham      830000.0   820000.0  553000.0
Miles          1090000.0   721500.0  555000.0
Alexkarbon     1036250.0   935000.0  560000.0
hockingstuart  1055000.0   935000.0  566000.0
Williams       1161000.0   735000.0  577000.0
Purplebricks    810000.0   927500.0  582500.0
Jellis         1505000.0  1052000.0  591750.0
Collins        1566500.0   930000.0  592000.0
Woodards       1261000.0   865500.0  600000.0
Eview           760000.0   597500.0  602000.0
Hodges         1500000.0  1105000.0  605500.0
O'Brien         737504.0   430000.0  618500.0
Moonee          806000.0   715000.0  622000.0
RW              870000.0   630500.0  625500.0
C21             836500.0   768750.0  625500.0
RT             1640000.0  1400000.0  630000.0
Philip         1035000.0   701000.0  636000.0
Chisholm       1520000.0   950000.0  640000.0
Fletchers      1390000.0  1238000.0  653000.0
Buxton         1323750.0  1030000.0  670000.0
Noel           1400500.0   990000.0  693250.0
Kay            2220000.0  1365000.0  695000.0
Cayzer         1505000.0  1450000.0  707500.0
Marshall       1975000.0  1408500.0  715000.0
Nick           2025000.0   780000.0  900000.0

In [44]:
movies_db = pd.read_csv('data/movies.csv')

In [47]:
movies_db['title'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 9742 entries, 0 to 9741
Series name: title
Non-Null Count  Dtype 
--------------  ----- 
9742 non-null   object
dtypes: object(1)
memory usage: 76.2+ KB


In [51]:
rat1_db = pd.read_csv('data/ratings1.csv')
rat1_db.nunique()

userId      274
movieId    6219
rating       10
dtype: int64

In [52]:
dates_db = pd.read_csv('data/dates.csv')
dates_db.nunique()


date    85043
dtype: int64

In [59]:
dates_db['date'] = pd.to_datetime(dates_db['date'], dayfirst=True)

In [65]:
dates_db['date'] = pd.to_datetime(dates_db['date'])
dates_db['date'] = dates_db['date'].dt.year
dates_db['date'].value_counts()

2000    10061
2017     8198
2007     7114
2016     6703
2015     6616
2018     6418
1996     6040
2005     5813
2012     4656
2008     4351
2009     4158
2006     4059
2003     4014
2001     3922
2002     3478
2004     3279
1999     2439
2010     2301
1997     1916
2011     1690
2013     1664
2014     1439
1998      507
Name: date, dtype: int64

In [67]:
rat2_db = pd.read_csv('data/ratings2.csv')
ratings = pd.concat([rat1_db, rat2_db])
display(ratings)

userId  movieId  rating
0           1        1     4.0
1           1        3     4.0
2           1        6     4.0
3           1       47     5.0
4           1       50     5.0
...       ...      ...     ...
60831     610   166534     4.0
60832     610   168248     5.0
60833     610   168250     5.0
60834     610   168252     5.0
60835     610   170875     3.0

[100837 rows x 3 columns]

In [69]:
ratings = pd.concat(
    [rat1_db, rat2_db],
    ignore_index=True
)
display(ratings)

userId  movieId  rating
0            1        1     4.0
1            1        3     4.0
2            1        6     4.0
3            1       47     5.0
4            1       50     5.0
...        ...      ...     ...
100832     610   166534     4.0
100833     610   168248     5.0
100834     610   168250     5.0
100835     610   168252     5.0
100836     610   170875     3.0

[100837 rows x 3 columns]

In [70]:
display('Число строк в таблице ratings: ', ratings.shape[0])
display('Число строк в таблице dates: ', dates_db.shape[0])
display(ratings.shape[0] == dates_db.shape[0])

'Число строк в таблице ratings: '

100837

'Число строк в таблице dates: '

100836

False

In [72]:
display(rat1_db.tail(1))
display(rat2_db.head(1))

userId  movieId  rating
40000     274     5621     2.0

userId  movieId  rating
0     274     5621     2.0

In [73]:
ratings = ratings.drop_duplicates(ignore_index=True)
display('Число строк в таблице ratings: ', ratings.shape[0])

'Число строк в таблице ratings: '

100836

In [74]:
ratings_dates = pd.concat([ratings, dates_db], axis=1)
display(ratings_dates.tail(7))

userId  movieId  rating  date
100829     610   164179     5.0  2017
100830     610   166528     4.0  2017
100831     610   166534     4.0  2017
100832     610   168248     5.0  2017
100833     610   168250     5.0  2017
100834     610   168252     5.0  2017
100835     610   170875     3.0  2017

import pandas as pd
import os

def concat_users_files(path):
    data = pd.DataFrame()
    file_names = os.listdir(path)
    file_names.sort()
    for file in file_names:
        tmp_data = pd.read_csv(path + '/' + file)
        data = pd.concat([data, tmp_data], axis=0, ignore_index=True)
    data = data.drop_duplicates()
    return data
    """
    Вам необходимо написать функцию concat_user_files(path), параметром которой является path - путь до директории. 
    Функция должна объединить информацию из предоставленных вам файлов в один DataFrame и вернуть его. 
    Не забудьте обновить индексы результирующей таблицы после объединения.
    Учтите тот момент, что в результате объединения могут возникнуть дубликаты, от которых необходимо будет избавиться. 
    """


if __name__ == '__main__':
    data = concat_users_files('./Root/users/')

In [76]:
movies=pd.read_csv('data/movies.csv')

In [77]:
joined_false = ratings_dates.join(
    movies,
    rsuffix='_right',
    how='left'
)
display(joined_false)

userId  movieId  rating  date  movieId_right  \
0            1        1     4.0  2000            1.0   
1            1        3     4.0  2000            2.0   
2            1        6     4.0  2000            3.0   
3            1       47     5.0  2000            4.0   
4            1       50     5.0  2000            5.0   
...        ...      ...     ...   ...            ...   
100831     610   166534     4.0  2017            NaN   
100832     610   168248     5.0  2017            NaN   
100833     610   168250     5.0  2017            NaN   
100834     610   168252     5.0  2017            NaN   
100835     610   170875     3.0  2017            NaN   

                                     title  \
0                         Toy Story (1995)   
1                           Jumanji (1995)   
2                  Grumpier Old Men (1995)   
3                 Waiting to Exhale (1995)   
4       Father of the Bride Part II (1995)   
...                                    ...   
100831                                 NaN   
100832                                 NaN   
100833                                 NaN   
100834                                 NaN   
100835                                 NaN   

                                             genres  
0       Adventure|Animation|Children|Comedy|Fantasy  
1                        Adventure|Children|Fantasy  
2                                    Comedy|Romance  
3                              Comedy|Drama|Romance  
4                                            Comedy  
...                                             ...  
100831                                          NaN  
100832                                          NaN  
100833                                          NaN  
100834                                          NaN  
100835                                          NaN  

[100836 rows x 7 columns]

In [78]:
joined = ratings_dates.join(
    movies.set_index('movieId'),
    on='movieId',
    how='left'
)
display(joined.head())

userId  movieId  rating  date                        title  \
0       1        1     4.0  2000             Toy Story (1995)   
1       1        3     4.0  2000      Grumpier Old Men (1995)   
2       1        6     4.0  2000                  Heat (1995)   
3       1       47     5.0  2000  Seven (a.k.a. Se7en) (1995)   
4       1       50     5.0  2000   Usual Suspects, The (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                               Comedy|Romance  
2                        Action|Crime|Thriller  
3                             Mystery|Thriller  
4                       Crime|Mystery|Thriller

In [79]:
merged = ratings_dates.merge(
    movies,
    on='movieId',
    how='left'
)
display(merged.head())

userId  movieId  rating  date                        title  \
0       1        1     4.0  2000             Toy Story (1995)   
1       1        3     4.0  2000      Grumpier Old Men (1995)   
2       1        6     4.0  2000                  Heat (1995)   
3       1       47     5.0  2000  Seven (a.k.a. Se7en) (1995)   
4       1       50     5.0  2000   Usual Suspects, The (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                               Comedy|Romance  
2                        Action|Crime|Thriller  
3                             Mystery|Thriller  
4                       Crime|Mystery|Thriller

import pandas as pd

items_df = pd.DataFrame({
'item_id': [417283, 849734, 132223, 573943, 19475, 3294095, 382043, 302948, 100132, 312394], 
'vendor': ['Samsung', 'LG', 'Apple', 'Apple', 'LG', 'Apple', 'Samsung', 'Samsung', 'LG', 'ZTE'],
'stock_count': [54, 33, 122, 18, 102, 43, 77, 143, 60, 19]
})

purchase_df = pd.DataFrame({
    'purchase_id': [101, 101, 101, 112, 121, 145, 145, 145, 145, 221],
    'item_id': [417283, 849734, 132223, 573943, 19475, 3294095, 382043, 302948, 103845, 100132], 
    'price': [13900, 5330, 38200, 49990, 9890, 33000, 67500, 34500, 89900, 11400]
})
"""
Сформируйте DataFrame merged, в котором в результате объединения
purchase_df и items_df останутся модели, которые учтены на складе и имели продажи. 
"""
merged = items_df.merge(purchase_df, how='inner', on='item_id')

"""
Найдите из таблицы merged суммарную выручку, которую можно было бы получить 
от продажи всех товаров, которые есть на складе. 
Результат занесите в переменную income.
"""

income = (merged['price'] * merged['stock_count']).sum()

PYTHON-12. Продвинутые методы работы с данными в Pandas  8. Закрепление знаний

In [133]:
import re 
def get_year_release(arg):
    #находим все слова по шаблону "(DDDD)"
    candidates = re.findall(r'\(\d{4}\)', arg) 
    # проверяем число вхождений
    if len(candidates) > 0:
        #если число вхождений больше 0,
	#очищаем строку от знаков "(" и ")"
        year = candidates[0].replace('(', '')
        year = year.replace(')', '')
        return int(year)
    else:
        #если год не указан, возвращаем None
        return None

In [141]:
ratings_movies=pd.read_csv('data/ratings_movies.csv')

In [142]:
ratings_movies['date'] = pd.to_datetime(ratings_movies['date'], dayfirst=True)

In [137]:
ratings_movies['year_release'].isnull().sum()

KeyError: 'year_release'

In [138]:

joined['year_release'] = joined['title'].apply(get_year_release)
joined.info()

TypeError: 'int' object is not subscriptable

In [139]:
joined['year_release'].isnull().sum()

TypeError: 'int' object is not subscriptable

In [111]:
mask = joined['year_release'] == 1999
joined[mask].groupby('title')['rating'].mean().sort_values()

title
Bloodsport: The Dark Kumite (1999)            0.5
Simon Sez (1999)                              1.0
Chill Factor (1999)                           1.0
Source, The (1999)                            1.0
Trippin' (1999)                               1.0
                                             ... 
Trailer Park Boys (1999)                      5.0
Larry David: Curb Your Enthusiasm (1999)      5.0
Sun Alley (Sonnenallee) (1999)                5.0
George Carlin: You Are All Diseased (1999)    5.0
Five Senses, The (1999)                       5.0
Name: rating, Length: 261, dtype: float64

genres
Action|Sci-Fi                        1.000000
Action|Adventure|Horror              1.500000
Action|Drama|Fantasy                 1.500000
Crime|Romance                        1.500000
Adventure|Comedy|Fantasy             1.833333
                                       ...   
Crime                                4.750000
Comedy|Musical                       5.000000
Animation|Drama|Fantasy|Mystery      5.000000
Adventure|Children|Comedy|Mystery    5.000000
Animation|Children|Mystery           5.000000
Name: rating, Length: 119, dtype: float64

In [113]:
joined.groupby('userId')['genres'].nunique().sort_values()

userId
214     13
85      13
245     13
494     15
578     15
      ... 
474    395
380    399
448    403
414    482
599    524
Name: genres, Length: 610, dtype: int64

In [114]:
joined.groupby('userId')['rating'].agg(
    ['count', 'mean']
).sort_values(by='count', ascending=False)

count      mean
userId                 
414      2698  3.391957
599      2478  2.642050
474      2108  3.398956
448      1864  2.847371
274      1346  3.235884
...       ...       ...
442        20  1.275000
569        20  4.000000
320        20  3.525000
576        20  3.100000
53         20  5.000000

[610 rows x 2 columns]

In [117]:
mask = joined['year_release'] == 2018
grouped = joined[mask].groupby('genres')['rating'].agg(
    ['mean', 'count']
)
grouped[grouped['count']>10].sort_values(
    by='mean',
    ascending=False
)

mean  count
genres                                  
Action|Adventure|Sci-Fi  3.928571     14
Action|Comedy|Sci-Fi     3.875000     12

In [119]:
joined['date'] = pd.to_datetime(joined['date'], dayfirst=True)
joined['year_rating'] = joined['date'].dt.year

In [126]:
joined.tail()

userId  movieId  rating                          date  \
100831     610   166534     4.0 1970-01-01 00:00:00.000002017   
100832     610   168248     5.0 1970-01-01 00:00:00.000002017   
100833     610   168250     5.0 1970-01-01 00:00:00.000002017   
100834     610   168252     5.0 1970-01-01 00:00:00.000002017   
100835     610   170875     3.0 1970-01-01 00:00:00.000002017   

                                 title                       genres  \
100831                    Split (2017)        Drama|Horror|Thriller   
100832   John Wick: Chapter Two (2017)        Action|Crime|Thriller   
100833                  Get Out (2017)                       Horror   
100834                    Logan (2017)                Action|Sci-Fi   
100835  The Fate of the Furious (2017)  Action|Crime|Drama|Thriller   

        year_release  year_rating  
100831        2017.0         1970  
100832        2017.0         1970  
100833        2017.0         1970  
100834        2017.0         1970  
100835        2017.0         1970

In [128]:
joined['date'] = pd.to_datetime(joined['date'])
joined['year_rating'] = joined['date'].dt.year
pivot = joined.pivot_table(
    index='year_rating',
    columns='genres',
    values='rating',
    aggfunc='mean'
)
display(pivot)

genres       (no genres listed)    Action  Action|Adventure  \
year_rating                                                   
1970                   3.489362  2.935484          3.706306   

genres       Action|Adventure|Animation  Action|Adventure|Animation|Children  \
year_rating                                                                    
1970                           3.583333                             3.410256   

genres       Action|Adventure|Animation|Children|Comedy  \
year_rating                                               
1970                                           3.808989   

genres       Action|Adventure|Animation|Children|Comedy|Fantasy  \
year_rating                                                       
1970                                                    3.7375    

genres       Action|Adventure|Animation|Children|Comedy|IMAX  \
year_rating                                                    
1970                                                3.483333   

genres       Action|Adventure|Animation|Children|Comedy|Romance  \
year_rating                                                       
1970                                                  2.607143    

genres       Action|Adventure|Animation|Children|Comedy|Sci-Fi  ...  \
year_rating                                                     ...   
1970                                                  2.973684  ...   

genres       Romance|Thriller  Romance|War  Romance|Western   Sci-Fi  \
year_rating                                                            
1970                 2.954545         2.75         3.083333  3.31338   

genres       Sci-Fi|IMAX  Sci-Fi|Thriller  Sci-Fi|Thriller|IMAX  Thriller  \
year_rating                                                                 
1970            3.993151         3.280172              3.708333  3.426752   

genres            War  Western  
year_rating                     
1970         3.555556  3.65894  

[1 rows x 951 columns]

In [140]:

joined2['year_release'] = joined2['title'].apply(get_year_release)
joined2.info()

NameError: name 'joined2' is not defined

In [143]:
ratings_movies.head()

Unnamed: 0  userId  movieId  rating                date  \
0           0       1        1     4.0 2000-07-30 18:45:03   
1           1       1        3     4.0 2000-07-30 18:20:47   
2           2       1        6     4.0 2000-07-30 18:37:04   
3           3       1       47     5.0 2000-07-30 19:03:35   
4           4       1       50     5.0 2000-07-30 18:48:51   

                         title                                       genres  
0             Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy  
1      Grumpier Old Men (1995)                               Comedy|Romance  
2                  Heat (1995)                        Action|Crime|Thriller  
3  Seven (a.k.a. Se7en) (1995)                             Mystery|Thriller  
4   Usual Suspects, The (1995)                       Crime|Mystery|Thriller

In [144]:
ratings_movies['year_rating']=ratings_movies['date'].dt.year

In [146]:
pivot = ratings_movies.pivot_table(
    index='year_rating',
    columns='genres',
    values='rating',
    aggfunc='mean'
)
display(pivot)

genres       (no genres listed)    Action  Action|Adventure  \
year_rating                                                   
1996                        NaN  2.730769          3.454545   
1997                        NaN  3.538462          4.150000   
1998                        NaN       NaN          4.200000   
1999                        NaN       NaN          4.000000   
2000                        NaN  2.588235          3.738462   
2001                        NaN  3.000000          3.500000   
2002                        NaN  2.750000          4.304348   
2003                        NaN  3.833333          3.277778   
2004                        NaN  2.700000          4.136364   
2005                        NaN  3.357143          3.413043   
2006                        NaN  2.166667          4.090909   
2007                        NaN  2.583333          3.666667   
2008                        NaN  3.000000          3.656250   
2009                        NaN       NaN          3.714286   
2010                        NaN  3.500000          3.437500   
2011                        NaN  4.500000          3.888889   
2012                        NaN  2.625000          3.900000   
2013                        NaN  2.500000          3.833333   
2014                        NaN       NaN          3.875000   
2015                   0.500000  3.100000          3.560976   
2016                   3.500000  3.545455          3.722222   
2017                   3.000000  2.637931          3.527778   
2018                   3.676471  2.588235          3.397436   

genres       Action|Adventure|Animation  Action|Adventure|Animation|Children  \
year_rating                                                                    
1996                                NaN                                  NaN   
1997                                NaN                                  NaN   
1998                                NaN                                  NaN   
1999                                NaN                                  NaN   
2000                                NaN                                  NaN   
2001                                NaN                                  NaN   
2002                                NaN                                  NaN   
2003                                NaN                                  NaN   
2004                                NaN                                  NaN   
2005                                NaN                                  NaN   
2006                                NaN                                  NaN   
2007                                NaN                                  NaN   
2008                                NaN                                  NaN   
2009                                NaN                                  NaN   
2010                                NaN                                  NaN   
2011                                NaN                                  NaN   
2012                                NaN                             4.000000   
2013                                NaN                             3.000000   
2014                              3.000                             3.500000   
2015                              3.800                             3.666667   
2016                              3.750                             3.833333   
2017                              3.375                             3.500000   
2018                              3.650                             3.166667   

genres       Action|Adventure|Animation|Children|Comedy  \
year_rating                                               
1996                                                NaN   
1997                                                NaN   
1998                                                NaN   
1999                                                NaN   
2000                                                NaN   
2001      

In [150]:
orders_db = pd.read_csv('data/orders.csv',sep=';')

In [151]:

products_db = pd.read_csv('data/products.csv',sep=';')

In [152]:
orders_db.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18 entries, 0 to 17
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Дата создания  18 non-null     object
 1   Order ID       18 non-null     int64 
 2   ID Покупателя  18 non-null     int64 
 3   Статус         18 non-null     object
 4   Оплачен        18 non-null     object
 5   Отменен        18 non-null     object
 6   Отгружен       18 non-null     object
 7   ID товара      18 non-null     int64 
 8   Количество     18 non-null     int64 
dtypes: int64(4), object(5)
memory usage: 1.4+ KB


In [153]:
products_db.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14 entries, 0 to 13
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Product_ID  14 non-null     int64 
 1   Name        14 non-null     object
 2   Price       14 non-null     int64 
 3   CURRENCY    14 non-null     object
dtypes: int64(2), object(2)
memory usage: 576.0+ bytes


In [154]:
orders_products = orders_db.merge(
    products_db, 
    left_on='ID товара',
    right_on='Product_ID',
    how='left')
orders_products.tail(1)['Order ID']

17    0
Name: Order ID, dtype: int64

In [157]:
orders_products.tail()

Дата создания  Order ID  ID Покупателя  \
13  08.11.2019 08:36:21         2              7   
14  08.11.2019 08:36:20         1              5   
15  08.11.2019 08:36:20         1              5   
16  08.11.2019 08:36:20         1              5   
17  01.01.2001 00:00:00         0              1   

                             Статус Оплачен Отменен Отгружен  ID товара  \
13                         Выполнен      Да     Нет       Да         47   
14  Оплачен, формируется к отправке      Да     Нет      Нет         71   
15  Оплачен, формируется к отправке      Да     Нет      Нет         86   
16  Оплачен, формируется к отправке      Да     Нет      Нет         51   
17  Оплачен, формируется к отправке      Да     Нет      Нет        666   

    Количество  Product_ID                      Name   Price CURRENCY  
13           1        47.0      Шатны Полосатый рейс  2999.0      RUR  
14           1        71.0       Платье Ночная Жизнь  7999.0      RUR  
15           1        86.0        Носки Простые, муж    45.0      RUR  
16           1        51.0  Платье Аленький цветочек  4999.0      RUR  
17           1         NaN                       NaN     NaN      NaN

In [161]:
orders_products['Статус']=='Отменён'

0     False
1     False
2     False
3     False
4     False
5     False
6      True
7     False
8     False
9     False
10    False
11    False
12    False
13    False
14    False
15    False
16    False
17    False
Name: Статус, dtype: bool

In [160]:
orders_products.head(10)

Дата создания  Order ID  ID Покупателя  \
0  09.11.2019 21:55:51         9             10   
1  09.11.2019 15:05:57         8              9   
2  09.11.2019 15:05:57         8              9   
3  09.11.2019 12:50:07         7              8   
4  09.11.2019 12:00:00         6              1   
5  09.11.2019 12:00:00         6              1   
6  08.11.2019 08:36:22         5              5   
7  08.11.2019 08:36:22         4              9   
8  08.11.2019 08:36:22         3              8   
9  08.11.2019 08:36:22         3              8   

                            Статус Оплачен Отменен Отгружен  ID товара  \
0         Принят, ожидается оплата     Нет     Нет      Нет        103   
1         Принят, ожидается оплата     Нет     Нет      Нет         86   
2         Принят, ожидается оплата     Нет     Нет      Нет        104   
3         Принят, ожидается оплата     Нет     Нет      Нет        104   
4         Принят, ожидается оплата     Нет     Нет      Нет        104   
5         Принят, ожидается оплата     Нет     Нет      Нет        103   
6                          Отменён     Нет      Да      Нет        124   
7         Принят, ожидается оплата     Нет     Нет       Да         91   
8  Оплачен, формируется к отправке      Да     Нет      Нет        103   
9  Оплачен, формируется к отправке      Да     Нет      Нет        104   

   Количество  Product_ID                   Name  Price CURRENCY  
0           5       103.0  Носки Подарочные, муж  199.0      RUR  
1         100        86.0     Носки Простые, муж   45.0      RUR  
2          10       104.0  Носки Подарочные, жен  249.0      RUR  
3           7       104.0  Носки Подарочные, жен  249.0      RUR  
4           5       104.0  Носки Подарочные, жен  249.0      RUR  
5           5       103.0  Носки Подарочные, муж  199.0      RUR  
6           1       124.0   Носки беговые Camino  999.0      RUR  
7           1        91.0     Носки Честные, муж   50.0      RUR  
8           3       103.0  Носки Подарочные, муж  199.0      RUR  
9           3       104.0  Носки Подарочные, жен  249.0      RUR

In [166]:
orders_products['Profit'] = orders_products['Price'] * orders_products['Количество'] 
orders_products[orders_products['Оплачен'] == 'Да'].groupby('ID Покупателя')['Profit'].sum().sort_values(ascending=False)

ID Покупателя
7    17096.0
5    13043.0
8     1344.0
1        0.0
Name: Profit, dtype: float64